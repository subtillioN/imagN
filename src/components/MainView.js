import { div, header, main, footer } from '@cycle/dom';
import { BaseView } from '../mvi/view';
import styles from '../styles/base.module.css';

export class MainView extends BaseView {
  constructor(sources) {
    super(sources);
    this.initializeStreams();
  }

  initializeStreams() {
    // Initialize component-specific streams
    this.imageConfig$ = this.sources.state.imageConfig$ || {};
    this.progress$ = this.sources.state.progress$ || {};
    this.results$ = this.sources.state.results$ || {};
  }

  renderHeader(state) {
    return header(`.${styles['app-header']}`, [
      div(`.${styles.flex}.${styles['items-center']}.${styles['gap-md']}`, [
        div(`.${styles.logo}`, 'ImagN'),
        div(`.${styles['nav-controls']}.${styles.flex}.${styles['gap-sm']}`, [
          div(`.${styles['button']}.${styles['button-primary']}`, 'New Project'),
          div(`.${styles.button}`, 'Gallery')
        ])
      ])
    ]);
  }

  renderContent(state) {
    return main(`.${styles['app-content']}`, [
      div(`.${styles.container}`, [
        div(`.${styles.workspace}`, [
          this.renderImageConfig(state),
          this.renderProgress(state),
          this.renderResults(state)
        ])
      ])
    ]);
  }

  renderImageConfig(state) {
    return div('.image-config', [
      // Image generation configuration UI will be implemented here
    ]);
  }

  renderProgress(state) {
    return div('.progress-section', [
      // Progress indicators will be implemented here
    ]);
  }

  renderResults(state) {
    return div('.results-section', [
      // Generated image results will be displayed here
    ]);
  }

  renderFooter(state) {
    return footer('.app-footer', [
      div('.footer-content', [
        // Footer content will be added here
      ])
    ]);
  }

  view(state) {
    return div('.app-container', [
      this.renderHeader(state),
      this.renderContent(state),
      this.renderFooter(state)
    ]);
  }
}