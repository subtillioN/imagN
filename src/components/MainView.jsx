import React from 'react';
import { BaseView } from '../mvi/view';
import styles from '../styles/base.module.css';

export class MainView extends BaseView {
  constructor(props) {
    super(props);
    this.sources = props.sources;
    this.state = {
      imageConfig: {},
      progress: {},
      results: {}
    };
    this.initializeStreams();
  }

  initializeStreams() {
    if (!this.sources || !this.sources.state) return;

    this.imageConfig$ = this.sources.state.imageConfig$ || {};
    this.progress$ = this.sources.state.progress$ || {};
    this.results$ = this.sources.state.results$ || {};

    if (this.imageConfig$.subscribe) {
      this.imageConfig$.subscribe(config => {
        this.setState({ imageConfig: config });
      });
    }

    if (this.progress$.subscribe) {
      this.progress$.subscribe(progress => {
        this.setState({ progress });
      });
    }

    if (this.results$.subscribe) {
      this.results$.subscribe(results => {
        this.setState({ results });
      });
    }
  }

  componentDidMount() {
    this.initializeStreams();
  }

  renderHeader(state) {
    return (
      <header className={styles['app-header']}>
        <div className={`${styles.flex} ${styles['items-center']} ${styles['gap-md']}`}>
          <div className={styles.logo}>ImagN</div>
          <nav className={styles['main-nav']}>
            <ul className={`${styles['nav-list']} ${styles.flex} ${styles['gap-md']}`}>
              <li role="listitem" aria-label="Image Workspace" className={`${styles['nav-item']} ${styles['active']}`}>Image Workspace</li>
              <li role="listitem" aria-label="Video Workspace" className={styles['nav-item']}>Video Workspace</li>
              <li role="listitem" aria-label="Node Editor" className={styles['nav-item']}>Node Editor</li>
              <li role="listitem" aria-label="Presets" className={styles['nav-item']}>Presets</li>
            </ul>
          </nav>
          <div className={`${styles['nav-controls']} ${styles.flex} ${styles['gap-sm']}`}>
            <div className={`${styles.button} ${styles['button-primary']}`}>New Project</div>
            <div className={styles.button}>Gallery</div>
          </div>
        </div>
      </header>
    );
  }

  renderContent(state) {
    return (
      <main className={styles['app-content']}>
        <div className={styles.container}>
          <div className={styles.workspace}>
            {this.renderProgress(state)}
            <div className={`${styles['workspace-panel']} ${styles['node-editor']}`}>
              <div className={styles['panel-header']}>Node Editor</div>
              <div className={styles['panel-content']}>Node canvas will be implemented here</div>
            </div>
            <div className={`${styles['workspace-panel']} ${styles['preview']}`}>
              <div className={styles['panel-header']}>Preview</div>
              <div className={styles['panel-content']}>Preview content will be displayed here</div>
            </div>
            <div className={`${styles['workspace-panel']} ${styles['properties']}`}>
              <div className={styles['panel-header']}>Properties</div>
              <div className={styles['panel-content']}>Node properties will be shown here</div>
            </div>
          </div>
        </div>
      </main>
    );
  }

  renderImageConfig(state) {
    return (
      <div className={styles['image-config']}>
        {/* Image generation configuration UI will be implemented here */}
      </div>
    );
  }

  renderProgress(state) {
    const { progress } = state;
    return (
      <div className={styles['progress-section']}>
        {progress && progress.status && (
          <div className={styles['progress-status']} role="status" aria-live="polite">{progress.status}</div>
        )}
      </div>
    );
  }

  renderResults(state) {
    const { results } = state;
    return (
      <div className={styles['results-section']}>
        {results.images && results.images.map((image, index) => (
          <div key={index} className={styles['result-image']}>
            <img src={image.url} alt={`Generated ${index + 1}`} />
          </div>
        ))}
      </div>
    );
  }

  renderFooter(state) {
    return (
      <footer className={styles['app-footer']}>
        <div className={styles['footer-content']}>
          <p>Â© 2023 ImagN. All rights reserved.</p>
        </div>
      </footer>
    );
  }

  view(state) {
    return (
      <div className={styles['app-container']}>
        {this.renderHeader(state)}
        {this.renderContent(state)}
        {this.renderFooter(state)}
      </div>
    );
  }
}