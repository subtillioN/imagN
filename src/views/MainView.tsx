import { Stream } from 'xstream';
import { h } from '@cycle/react';
import { MainViewState } from '../types/state';
import React from 'react';
import { PresetList } from './PresetList';
import { ProjectList } from './ProjectList';
import { Tabs, Tab } from '@mui/material';
import { TagFilter } from './TagFilter';
import { NewProjectDialog } from './dialogs/NewProjectDialog';
import { LoadProjectDialog } from './dialogs/LoadProjectDialog';
import { TagManagementDialog } from './dialogs/TagManagementDialog';
import { Snackbar } from './Snackbar';

export function MainView(state$: Stream<MainViewState>): Stream<JSX.Element> {
  return state$.map(state => 
    h('div', { className: 'main-view' }, [
      // Tabs
      h(Tabs, {
        sel: 'tabs',
        value: state.currentTab,
        onChange: (_, value) => value,
      }, [
        h(Tab, { label: 'Presets', value: 'presets' }),
        h(Tab, { label: 'Projects', value: 'projects' })
      ]),

      // Filters
      h(TagFilter, {
        selectedCategories: Array.from(state.filters.selectedCategories),
        selectedTypes: Array.from(state.filters.selectedTypes),
        selectedTags: Array.from(state.filters.selectedTags),
        tagSearchQuery: state.filters.tagSearchQuery,
        availableCategories: Array.from(state.availableCategories),
        availableTypes: Array.from(state.availableTypes),
        availableTags: Array.from(state.availableTags)
      }),

      // Content
      state.currentTab === 'presets' ?
        h(PresetList, {
          systemPresets: state.filteredSystemPresets,
          userPresets: state.filteredUserPresets,
          selectedPresetId: state.selectedProjectId,
          isLoading: state.isLoadingPresets
        }) :
        h(ProjectList, {
          projects: state.projects,
          selectedProjectId: state.selectedProjectId,
          isLoading: state.isLoadingProjects
        }),

      // Dialogs
      h(NewProjectDialog, {
        open: state.newProjectDialog.isOpen,
        systemPresets: state.systemPresets,
        userPresets: state.userPresets
      }),

      h(LoadProjectDialog, {
        open: state.loadProjectDialog.isOpen,
        projects: state.projects
      }),

      h(TagManagementDialog, {
        open: state.tagManagementDialog.isOpen,
        presetId: state.tagManagementDialog.presetId,
        availableTags: Array.from(state.availableTags)
      }),

      // Notifications
      h(Snackbar, {
        open: state.snackbar.isOpen,
        message: state.snackbar.message,
        severity: state.snackbar.severity,
        autoHideDuration: state.snackbar.autoHideDuration
      })
    ])
  );
} 