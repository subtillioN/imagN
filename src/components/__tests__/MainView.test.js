import { MainView } from '../MainView';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import subject from 'callbag-subject';

describe('MainView Component', () => {
  const createMockStream = () => {
    const subj = subject();
    return {
      subscribe: (callback) => {
        subject(0, callback);
        return () => subject(2);
      }
    };
  };

  const createSources = () => ({
    state: {
      imageConfig$: createMockStream(),
      progress$: createMockStream(),
      results$: createMockStream()
    }
  });

  it('should render correctly and match snapshot', () => {
    const sources = createSources();
    const mainView = new MainView(sources);
    const { container } = render(mainView.view({}));
    
    expect(container).toMatchSnapshot();
  });

  it('should render all main sections', () => {
    const sources = createSources();
    const mainView = new MainView(sources);
    const { container } = render(mainView.view({}));

    expect(container.querySelector('.app-header')).toBeInTheDocument();
    expect(container.querySelector('.app-content')).toBeInTheDocument();
    expect(container.querySelector('.app-footer')).toBeInTheDocument();
    expect(container.querySelector('.workspace')).toBeInTheDocument();
  });

  it('should update state when streams emit new values', () => {
    const sources = createSources();
    const mainView = new MainView(sources);
    
    render(mainView.view({ progress: { status: 'Processing' } }));
    
    expect(screen.getByText('Processing')).toBeInTheDocument();
  });

  it('should render navigation items correctly', () => {
    const sources = createSources();
    const mainView = new MainView(sources);
    render(mainView.view({}));

    expect(screen.getByRole('listitem', { name: 'Image Workspace' })).toBeInTheDocument();
    expect(screen.getByRole('listitem', { name: 'Video Workspace' })).toBeInTheDocument();
    expect(screen.getByRole('listitem', { name: 'Node Editor' })).toBeInTheDocument();
    expect(screen.getByRole('listitem', { name: 'Presets' })).toBeInTheDocument();
  });
});