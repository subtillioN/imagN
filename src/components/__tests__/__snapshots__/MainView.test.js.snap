// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainView Component should render correctly and match snapshot 1`] = `
<div>
  <div
    class="app-container"
  >
    <header
      class="app-header"
    >
      <div
        class="flex items-center gap-md"
      >
        <div
          class="logo"
        >
          ImagN
        </div>
        <nav
          class="main-nav"
        >
          <ul
            class="nav-list flex gap-md"
          >
            <li
              aria-label="Image Workspace"
              class="nav-item active"
              role="listitem"
            >
              Image Workspace
            </li>
            <li
              aria-label="Video Workspace"
              class="nav-item"
              role="listitem"
            >
              Video Workspace
            </li>
            <li
              aria-label="Node Editor"
              class="nav-item"
              role="listitem"
            >
              Node Editor
            </li>
            <li
              aria-label="Presets"
              class="nav-item"
              role="listitem"
            >
              Presets
            </li>
          </ul>
        </nav>
        <div
          class="nav-controls flex gap-sm"
        >
          <div
            class="button button-primary"
          >
            New Project
          </div>
          <div
            class="button"
          >
            Gallery
          </div>
        </div>
      </div>
    </header>
    <main
      class="app-content"
    >
      <div
        class="container"
      >
        <div
          class="workspace"
        >
          <div
            class="progress-section"
          />
          <div
            class="workspace-panel node-editor"
          >
            <div
              class="panel-header"
            >
              Node Editor
            </div>
            <div
              class="panel-content"
            >
              Node canvas will be implemented here
            </div>
          </div>
          <div
            class="workspace-panel preview"
          >
            <div
              class="panel-header"
            >
              Preview
            </div>
            <div
              class="panel-content"
            >
              Preview content will be displayed here
            </div>
          </div>
          <div
            class="workspace-panel properties"
          >
            <div
              class="panel-header"
            >
              Properties
            </div>
            <div
              class="panel-content"
            >
              Node properties will be shown here
            </div>
          </div>
        </div>
      </div>
    </main>
    <footer
      class="app-footer"
    >
      <div
        class="footer-content"
      >
        <p>
          Â© 2023 ImagN. All rights reserved.
        </p>
      </div>
    </footer>
  </div>
</div>
`;
